### PR TITLE
New Converter

### LINKED ISSUE #406: Can't convert a graph with Flatten as the first node
If you try to secure a model via the converter where flatten is the first node, it will crash

```python
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(512, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(x_train, y_train, epochs=1)
model.evaluate(x_test, y_test)

secret_model = tfe.secure_model(model)
prediction = secret_model.private_predict(provide_input())
```

```
Traceback (most recent call last):
  File "ignite.py", line 41, in <module>
    secret_model = tfe.secure_model(model)
  File "/home/bendecoste/tf-encrypted/tf_encrypted/private_model.py", line 70, in secure_model
    y = c.convert(remove_training_nodes(graph_def), tfe.convert.register(), 'input-provider', inputs)
  File "/home/bendecoste/tf-encrypted/tf_encrypted/convert/convert.py", line 92, in convert
    outs = register[special_op_dict[s]['op']](self, node, input_list)
  File "/home/bendecoste/tf-encrypted/tf_encrypted/convert/register.py", line 421, in flatten
    input = converter.outputs[inputs[0]]
IndexError: list index out of range
```

### PR DESCRIPTION
**Pull request by jvmancuso**
Fixes #406 
Related #297

#### Additions
- "Special op" conversions can now use pre-specified intermediary nodes during TFE graph construction.  For example, if a subgraph contains a non-leaf Conv2D op (i.e., an op that is not a direct input or output node of the subgraph), a conversion for the subgraph can extract attributes from the GraphDef for that Conv2D op.  In particular, this allows us to convert Keras layers in the future.
- Accordingly, this also gives the option to fuse higher-level ops and layers through the converter, as mentioned in #297.  One would do so by collecting the TF ops under some higher level scope, properly registering that scope inside `_REGISTERED_SPECIAL_OPS` in convert.py, and then writing a specialized conversion function in register.py to specify how the ops should be constructed in TFE.
- The converter now searches for scopes and ops with regex, making it slightly more robust.

#### Fixes
- Previously, the converter scanned a full GraphDef 5 times.  It now does so 3 times -- once for compiling special ops, once for pruning those out, and once for constructing the TFE graph.  This can likely be improved further if needed.
- Moved `Converter` into the `tf_encrypted` namespace
- The test for converting `tf.keras.layers.Flatten` included a Conv2D layer, which has been removed.
- Flatten can now be the first op in a graph. (#406)
- test_convert.py is now seeded.
- The `expand_dims` method for PondTensor was missing an `axis` kwarg

#### Miscellaneous:
- Had a few flake problems on my machine, removed some type hints accordingly.
- Improved a few error messages.
- Renamed `register` to `registry` in `tf_encrypted/convert/register.py`

TODO:
- [x] Finish cleaning `required_batch_to_paddings` conversion
- [x] Comment converter functions
- [x] Finish writing `keras_conv2d` conversion test case
- [x] Write a how-to for adding new conversions with test cases
- [x] Maintain a list of scopes reserved for special ops (should be codegen'd from the keys of `REGISTERED_SPECOPS`)

### CODE REVIEW

File: tf_encrypted/convert/register.py [hunk 1]
```diff
@@ -544,8 +572,10 @@ def inputter_fn():
     return x_public
 
 
-def nodef_to_private_pond(converter, x):
+def nodef_to_private_pond(converter, x, int_warn=False, int_err=False):
```
Comment (justin1121) [inline tf_encrypted/convert/register.py]: What's the idea behind the `int_warn` and `int_err`? I don't see them being used anywhere.

File: tf_encrypted/convert/convert.py [hunk 2]
```diff
@@ -110,98 +115,138 @@ class InvalidArgumentError(Exception):
...
+            if input in output_leaves:
+                output_leaves.remove(input)
+    seen = set()
+    adder = seen.add
+    input_leaves = [x for x in input_leaves if not (x in seen or adder(x))]
+
+    return input_leaves, output_leaves
+
+
+def get_intermediaries(specop_scope, subscope_map):
```
Comment (justin1121) [inline tf_encrypted/convert/convert.py]: Can you explain what you mean by an intermediary here in a comment? Its not immediately obvious to me.

File: tf_encrypted/convert/register.py (hunk 1 above)
Comment (jvmancuso) [reply tf_encrypted/convert/register.py]: the idea was that most of the time you'll have floats or want to convert to those before defining an input, so if you end up with ints at that point you might have a bug/unexpected behavior.  I think I deleted the code that used it :(

Comment (justin1121) [reply tf_encrypted/convert/register.py]: Seems reasonable to me!

File: tf_encrypted/convert/convert.py (hunk 2 above)
Comment (jvmancuso) [reply tf_encrypted/convert/convert.py]: `An intermediary op for a subgraph is an op that is neither an input nor an output to that subgraph.` would that work?

File: tf_encrypted/convert/register.py (hunk 1 above)
Comment (mortendahl) [reply tf_encrypted/convert/register.py]: I agree, if it's just for setting the log level then it should probably be done elsewhere

File: tf_encrypted/convert/register.py [hunk 3]
```diff
@@ -555,9 +585,13 @@ def nodef_to_private_pond(converter, x):
         elif dtype == tf.float64:
             nums = x.attr["value"].tensor.float_val
         elif dtype == tf.int32:
+            if int_err:
```
Comment (mortendahl) [inline tf_encrypted/convert/register.py]: What's the reason for warning on this?

File: tf_encrypted/convert/convert.py (hunk 2 above)
Comment (mortendahl) [reply tf_encrypted/convert/convert.py]: I'd call that on internal op

File: tf_encrypted/convert/register.py (hunk 3 above)
Comment (jvmancuso) [reply tf_encrypted/convert/register.py]: see above comment from @justin1121, will fix in a later commit

Comment (mortendahl) [reply tf_encrypted/convert/register.py]: Answered above

File: tf_encrypted/convert/convert.py (hunk 2 above)
Comment (justin1121) [reply tf_encrypted/convert/convert.py]: Yep sounds good!

**Comment by mortendahl** [comment]
Overall looks good. Anything in particular you'd like a more thorough review on?

**Comment by jvmancuso** [comment]
As long as the general design looks good, I think I'm okay.  I'll carry on with the TODOs above and ping for final approval when ready.

File: tf_encrypted/protocol/pond.py [hunk 4]
```diff
@@ -490,7 +490,7 @@ def define_output(
             player = get_config().get_player(player)
         assert isinstance(player, Player)
 
-        def helper(x: Union["PondPrivateTensor", "PondMasterTensor"]) -> tf.Tensor:
+        def helper(x: Union["PondPrivateTensor", "PondMaskedTensor"]) -> tf.Tensor:
```
Comment (mortendahl) [inline tf_encrypted/protocol/pond.py]: good catch!

**Comment by jvmancuso** [comment]
@justin1121 @mortendahl @yanndupis this should be ready for final approval/merging

File: Makefile [hunk 5]
```diff
@@ -63,7 +63,16 @@ test: lint pythoncheck
 	pytest -n 8 -x -m convert_maxpool
 
 
-lint: pythoncheck
+CONVERT_DIR=$(CURRENT_DIR)/tf_encrypted/convert
```
Comment (jvmancuso) [inline Makefile]: this can probably be done more cleanly, but I'm new to make so would appreciate a pointer

File: Makefile [hunk 6]
```diff
@@ -63,7 +63,16 @@ test: lint pythoncheck
...
+BUILD_RESERVED_SCOPES=$(CONVERT_DIR)/specops.yaml
+$(BUILD_RESERVED_SCOPES): pythoncheck
+	python -m tf_encrypted.convert.gen.generate_reserved_scopes
+
+BUILD_CONVERTER_README=$(CONVERT_DIR)/gen/readme_template.md
+$(BUILD_CONVERTER_README): $(BUILD_RESERVED_SCOPES) pythoncheck
```
Comment (ianlivingstone) [inline Makefile]: 💯

File: Makefile (hunk 5 above)
Comment (ianlivingstone) [reply Makefile]: You run make from the project root so you shouldn't need the `$(CURRENT_DIR)` :) Otherwise, looks great!

**Pull request merged**